File: platforms/android/java/app/msime/client/ChineseOutputPolicy.java
package app.msime.client;

/** Host-boundary Simplified/Traditional output policy without Android dependencies. */
public final class ChineseOutputPolicy {
    @FunctionalInterface
    public interface Converter {
        String convert(String text);
    }

    private ChineseOutputPolicy() {}

    public static boolean applies(boolean dedicatedEnglish, int scheme, String localMode) {
        return !dedicatedEnglish && scheme != 3 && !"temporary_japanese".equals(localMode);
    }

    public static String output(String text, boolean traditional, boolean applies,
                                Converter converter) {
        if (!traditional || !applies || text.isEmpty()) return text;
        try {
            String converted = converter.convert(text);
            return converted == null ? text : converted;
        } catch (RuntimeException | LinkageError error) {
            return text;
        }
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/input/ChineseOutputPolicy.ts
/**
 * Simplified/Traditional output boundary, ported from
 * platforms/android/java/app/msime/client/ChineseOutputPolicy.java.
 *
 * Scheme 3 is Japanese, which has nothing to convert. A converter that fails must not lose the text,
 * so any error falls back to the original.
 */
export type ChineseConverter = (text: string) => string | null;

export class ChineseOutputPolicy {
  static applies(dedicatedEnglish: boolean, scheme: number, localMode: string): boolean {
    return !dedicatedEnglish && scheme !== 3 && localMode !== 'temporary_japanese';
  }

  static output(text: string, traditional: boolean, applies: boolean,
                converter: ChineseConverter): string {
    if (!traditional || !applies || text.length === 0) {
      return text;
    }
    try {
      const converted: string | null = converter(text);
      return converted === null ? text : converted;
    } catch (error) {
      return text;
    }
  }
}
